File: scripts/build-sources.js
#!/usr/bin/env node
"use strict";

const fs = require("fs");
const path = require("path");

const ROOT = path.resolve(__dirname, "..");
const UA_SPLIT = /##\s+Українська\s+Версія/i;
const EN_VERSION = /##\s+English\s+Version\s*/i;

const MODULE_MAP = {
  "characters-index.json": { indexGlobal: "CHARACTERS_LIBRARY_INDEX", mdGlobal: "__CHARACTERS_MD__" },
  "spells-index.json": { indexGlobal: "SPELL_LIBRARY_INDEX", mdGlobal: "__SPELL_MD__" },
  "monsters-index.json": { indexGlobal: "MONSTERS_LIBRARY_INDEX", mdGlobal: "__MONSTERS_MD__" },
  "npc-index.json": { indexGlobal: "NPC_LIBRARY_INDEX", mdGlobal: "__NPC_MD__" },
  "items-index.json": { indexGlobal: "ITEMS_LIBRARY_INDEX", mdGlobal: "__ITEMS_MD__" },
  "maps-index.json": { indexGlobal: "MAPS_LIBRARY_INDEX", mdGlobal: "__MAPS_MD__" },
  "dm-script-index.json": { indexGlobal: "DMSCRIPT_LIBRARY_INDEX", mdGlobal: "__DMSCRIPT_MD__" },
};

function walkDir(dir, out) {
  if (!fs.existsSync(dir)) return;
  for (const name of fs.readdirSync(dir)) {
    const full = path.join(dir, name);
    if (fs.statSync(full).isDirectory()) walkDir(full, out);
    else out.push(full);
  }
}

function findIndexFiles() {
  const files = [];
  walkDir(path.join(ROOT, "Shared"), files);
  walkDir(path.join(ROOT, "scenarios"), files);
  return files.filter(function (f) {
    const base = path.basename(f);
    return base.endsWith("-index.json") && MODULE_MAP[base];
  });
}

function slugFromIndexEntry(entry) {
  return String(entry)
    .replace(/^.*\//, "")
    .replace(/\.(en|ua)\.md$/i, "")
    .replace(/\.md$/i, "");
}

function readIndex(indexPath) {
  const raw = JSON.parse(fs.readFileSync(indexPath, "utf8"));
  if (!Array.isArray(raw)) throw new Error("Index must be array: " + indexPath);
  return raw.map(slugFromIndexEntry);
}

function writeIndex(indexPath, slugs) {
  fs.writeFileSync(indexPath, JSON.stringify(slugs, null, 2) + "\n", "utf8");
}

function escapeJsString(s) {
  return JSON.stringify(s);
}

function splitLegacyFile(filePath) {
  const content = fs.readFileSync(filePath, "utf8");
  if (!UA_SPLIT.test(content) && !EN_VERSION.test(content)) {
    return null;
  }
  const titleMatch = content.match(/^#\s+(.+)$/m);
  let enTitle = path.basename(filePath, ".md");
  let uaTitle = enTitle;
  if (titleMatch) {
    const parts = titleMatch[1].split(" / ");
    enTitle = parts[0].trim();
    uaTitle = parts.length > 1 ? parts.slice(1).join(" / ").trim() : enTitle;
  }

  const uaParts = content.split(UA_SPLIT);
  const beforeUa = uaParts[0];
  const uaBody = uaParts.length > 1 ? uaParts[1].trim() : "";

  const afterTitle = beforeUa.replace(/^#\s+.+\n?/, "").trim();
  const enMatch = afterTitle.match(EN_VERSION);
  let preamble = "";
  let enBody = "";
  if (enMatch) {
    const idx = afterTitle.search(EN_VERSION);
    preamble = afterTitle.slice(0, idx).trim();
    enBody = afterTitle.slice(idx + enMatch[0].length).trim();
  } else {
    enBody = afterTitle;
  }

  const preambleBlock = preamble ? preamble + "\n\n---\n\n" : "";
  const enFile = "# " + enTitle + "\n\n" + preambleBlock + enBody + "\n";
  const uaFile = uaBody
    ? "# " + uaTitle + "\n\n" + preambleBlock + uaBody + "\n"
    : null;

  return { enFile: enFile, uaFile: uaFile };
}

function splitLegacy() {
  const dirs = [];
  walkDir(path.join(ROOT, "Shared"), dirs);
  walkDir(path.join(ROOT, "scenarios"), dirs);
  const legacy = dirs.filter(function (f) {
    if (!f.endsWith(".md")) return false;
    if (f.endsWith(".en.md") || f.endsWith(".ua.md")) return false;
    if (path.basename(f) === "README.md") return false;
    const content = fs.readFileSync(f, "utf8");
    return UA_SPLIT.test(content) || EN_VERSION.test(content);
  });

  let split = 0;
  legacy.forEach(function (filePath) {
    const dir = path.dirname(filePath);
    const slug = path.basename(filePath, ".md");
    const enPath = path.join(dir, slug + ".en.md");
    const uaPath = path.join(dir, slug + ".ua.md");
    if (fs.existsSync(enPath)) {
      console.log("skip (en exists): " + path.relative(ROOT, filePath));
      return;
    }
    const result = splitLegacyFile(filePath);
    if (!result) return;
    fs.writeFileSync(enPath, result.enFile, "utf8");
    if (result.uaFile) fs.writeFileSync(uaPath, result.uaFile, "utf8");
    fs.unlinkSync(filePath);
    console.log("split: " + path.relative(ROOT, filePath));
    split += 1;
  });
  console.log("Split " + split + " legacy file(s).");
}

function validate() {
  let errors = 0;
  let warnings = 0;
  findIndexFiles().forEach(function (indexPath) {
    const dir = path.dirname(indexPath);
    const slugs = readIndex(indexPath);
    normalized.forEach(function (slug) {
      const enPath = path.join(dir, slug + ".en.md");
      const uaPath = path.join(dir, slug + ".ua.md");
      if (!fs.existsSync(enPath)) {
        console.error("MISSING EN: " + path.relative(ROOT, enPath));
        errors += 1;
      }
      if (!fs.existsSync(uaPath)) {
        console.warn("WARN missing UA: " + path.relative(ROOT, uaPath));
        warnings += 1;
      }
    });
  });
  console.log("Validate: " + errors + " error(s), " + warnings + " warning(s).");
  if (errors) process.exit(1);
}

function buildSources() {
  findIndexFiles().forEach(function (indexPath) {
    const base = path.basename(indexPath);
    const mod = MODULE_MAP[base];
    const dir = path.dirname(indexPath);
    const slugs = readIndex(indexPath);
    writeIndex(indexPath, slugs);

    const sourcesName = base.replace("-index.json", "-sources.js");
    const indexJsName = base.replace(".json", ".js");
    const sourcesPath = path.join(dir, sourcesName);
    const indexJsPath = path.join(dir, indexJsName);

    let sources = "window." + mod.mdGlobal + " = window." + mod.mdGlobal + " || {};\n";
    slugs.forEach(function (slug) {
      const enPath = path.join(dir, slug + ".en.md");
      const uaPath = path.join(dir, slug + ".ua.md");
      if (!fs.existsSync(enPath)) {
        console.error("Skip sources for missing EN: " + enPath);
        return;
      }
      const en = fs.readFileSync(enPath, "utf8");
      const ua = fs.existsSync(uaPath) ? fs.readFileSync(uaPath, "utf8") : "";
      sources += "window." + mod.mdGlobal + '["' + slug + '"] = { en: ' + escapeJsString(en);
      if (ua) sources += ", ua: " + escapeJsString(ua);
      sources += " };\n";
    });

    const indexJs = "window." + mod.indexGlobal + " = " + JSON.stringify(slugs) + ";\n";

    fs.writeFileSync(sourcesPath, sources, "utf8");
    fs.writeFileSync(indexJsPath, indexJs, "utf8");
    console.log("built: " + path.relative(ROOT, sourcesPath) + ", " + path.relative(ROOT, indexJsPath));
  });
}

const args = process.argv.slice(2);
if (args.includes("--split-legacy")) {
  splitLegacy();
}
if (!args.includes("--validate-only")) {
  buildSources();
}
if (args.includes("--validate") || args.includes("--validate-only")) {
  validate();
}

File: scripts/build-sources.py
#!/usr/bin/env python3
"""Validate content indexes and spell closure; split legacy bilingual markdown."""

import json
import os
import re
import shutil
import subprocess
import sys

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
UA_SPLIT = re.compile(r"##\s+Українська\s+Версія", re.I)
EN_VERSION = re.compile(r"##\s+English\s+Version\s*", re.I)

INDEX_BASENAMES = {
    "characters-index.json",
    "spells-index.json",
    "monsters-index.json",
    "npc-index.json",
    "items-index.json",
    "maps-index.json",
    "dm-script-index.json",
}


def walk_md_dirs():
    for base in (os.path.join(ROOT, "Shared"), os.path.join(ROOT, "scenarios")):
        for dirpath, _, filenames in os.walk(base):
            for name in filenames:
                yield os.path.join(dirpath, name)


def find_index_files():
    for path in walk_md_dirs():
        if os.path.basename(path) in INDEX_BASENAMES:
            yield path


def slug_from_index_entry(entry):
    s = str(entry).replace("\\", "/")
    s = s.split("/")[-1]
    s = re.sub(r"\.(en|ua)\.md$", "", s, flags=re.I)
    s = re.sub(r"\.md$", "", s, flags=re.I)
    return s


def read_index(index_path):
    with open(index_path, encoding="utf-8") as f:
        raw = json.load(f)
    if not isinstance(raw, list):
        raise ValueError("Index must be array: " + index_path)
    return [slug_from_index_entry(x) for x in raw]


def split_legacy_file(content, default_slug):
    if not UA_SPLIT.search(content) and not EN_VERSION.search(content):
        return None
    title_match = re.search(r"^#\s+(.+)$", content, re.M)
    en_title = default_slug
    ua_title = default_slug
    if title_match:
        parts = title_match.group(1).split(" / ")
        en_title = parts[0].strip()
        ua_title = " / ".join(parts[1:]).strip() if len(parts) > 1 else en_title

    ua_parts = UA_SPLIT.split(content, maxsplit=1)
    before_ua = ua_parts[0]
    ua_body = ua_parts[1].strip() if len(ua_parts) > 1 else ""

    after_title = re.sub(r"^#\s+.+\n?", "", before_ua, count=1).strip()
    en_match = EN_VERSION.search(after_title)
    preamble = ""
    en_body = ""
    if en_match:
        idx = en_match.start()
        preamble = after_title[:idx].strip()
        en_body = after_title[idx + len(en_match.group(0)) :].strip()
    else:
        en_body = after_title

    preamble_block = (preamble + "\n\n---\n\n") if preamble else ""
    en_file = "# " + en_title + "\n\n" + preamble_block + en_body + "\n"
    ua_file = None
    if ua_body:
        ua_file = "# " + ua_title + "\n\n" + preamble_block + ua_body + "\n"
    return en_file, ua_file


def split_legacy():
    count = 0
    for file_path in walk_md_dirs():
        if not file_path.endswith(".md"):
            continue
        if file_path.endswith(".en.md") or file_path.endswith(".ua.md"):
            continue
        if os.path.basename(file_path) == "README.md":
            continue
        with open(file_path, encoding="utf-8") as f:
            content = f.read()
        if not UA_SPLIT.search(content) and not EN_VERSION.search(content):
            continue
        dir_name = os.path.dirname(file_path)
        slug = os.path.splitext(os.path.basename(file_path))[0]
        en_path = os.path.join(dir_name, slug + ".en.md")
        ua_path = os.path.join(dir_name, slug + ".ua.md")
        if os.path.exists(en_path):
            print("skip (en exists):", os.path.relpath(file_path, ROOT))
            continue
        result = split_legacy_file(content, slug)
        if not result:
            continue
        en_file, ua_file = result
        with open(en_path, "w", encoding="utf-8") as f:
            f.write(en_file)
        if ua_file:
            with open(ua_path, "w", encoding="utf-8") as f:
                f.write(ua_file)
        os.remove(file_path)
        print("split:", os.path.relpath(file_path, ROOT))
        count += 1
    print("Split", count, "legacy file(s).")


def validate():
    errors = 0
    warnings = 0
    for index_path in find_index_files():
        dir_name = os.path.dirname(index_path)
        for slug in read_index(index_path):
            en_path = os.path.join(dir_name, slug + ".en.md")
            ua_path = os.path.join(dir_name, slug + ".ua.md")
            if not os.path.exists(en_path):
                print("MISSING EN:", os.path.relpath(en_path, ROOT))
                errors += 1
            if not os.path.exists(ua_path):
                print("WARN missing UA:", os.path.relpath(ua_path, ROOT))
                warnings += 1
    print("Validate:", errors, "error(s),", warnings, "warning(s).")
    if errors:
        sys.exit(1)


def run_spell_closure():
    node = shutil.which("node")
    if not node:
        print("Spell closure validation requires Node.js. Install Node or run: node scripts/build-sources.js")
        sys.exit(1)
    script = os.path.join(os.path.dirname(os.path.abspath(__file__)), "spell-cli.js")
    result = subprocess.run([node, script, "validate"], cwd=ROOT)
    if result.returncode:
        sys.exit(result.returncode)


def run_spell_sync():
    node = shutil.which("node")
    if not node:
        print("Sync spells requires Node.js. Install Node or run: node scripts/build-sources.js --sync-spells")
        sys.exit(1)
    script = os.path.join(os.path.dirname(os.path.abspath(__file__)), "spell-cli.js")
    result = subprocess.run([node, script, "sync"], cwd=ROOT)
    if result.returncode:
        sys.exit(result.returncode)


def main():
    args = sys.argv[1:]
    if "--split-legacy" in args:
        split_legacy()
    if "--sync-spells" in args:
        run_spell_sync()
    only_split_legacy = len(args) == 1 and args[0] == "--split-legacy"
    if not only_split_legacy:
        validate()
        run_spell_closure()


if __name__ == "__main__":
    main()
